tests/src/test/scala/whisk/http/PoolingRestClientTests.scala

```diff
@@ -0,0 +1,154 @@
+    await(poolingRestClient.request0(Future.successful(httpRequest))) shouldBe httpResponse
+  }
+
+  it should "return payload from a request" in {
+    val httpResponse = HttpResponse(entity = HttpEntity(ContentTypes.`application/json`, JsObject().compactPrint))
+    val httpRequest = HttpRequest()
+    val poolingRestClient = new PoolingRestClient("https", "host", 443, 1, Some(testFlow(httpResponse)))
+
+    await(poolingRestClient.request0(Future.successful(httpRequest))) shouldBe httpResponse
+  }
+
+  it should "send headers when making a request" in {
+    val httpResponse = HttpResponse(entity = HttpEntity(ContentTypes.`application/json`, JsObject().compactPrint))
+    val httpRequest = HttpRequest(headers = List(RawHeader("key", "value")))
+    val poolingRestClient = new PoolingRestClient("https", "host", 443, 1, Some(testFlow(httpResponse, httpRequest)))
+
+    await(poolingRestClient.request0(Future.successful(httpRequest))) shouldBe httpResponse
+  }
+
+  it should "send uri when making a request" in {
+    val httpResponse = HttpResponse(entity = HttpEntity(ContentTypes.`application/json`, JsObject().compactPrint))
+    val httpRequest = HttpRequest(uri = Uri("/some/where"))
+    val poolingRestClient = new PoolingRestClient("https", "host", 443, 1, Some(testFlow(httpResponse, httpRequest)))
+
+    await(poolingRestClient.request0(Future.successful(httpRequest))) shouldBe httpResponse
+  }
+
+  it should "send a payload when making a request" in {
+    val httpResponse = HttpResponse(entity = HttpEntity(ContentTypes.`application/json`, JsObject().compactPrint))
+    val httpRequest = HttpRequest(POST, entity = HttpEntity(ContentTypes.`text/plain(UTF-8)`, "payload"))
+    val poolingRestClient = new PoolingRestClient("https", "host", 443, 1, Some(testFlow(httpResponse, httpRequest)))
+
+    await(poolingRestClient.request0(Future.successful(httpRequest))) shouldBe httpResponse
+  }
+
+  it should "return JSON when making a request" in {
+    val httpResponse = HttpResponse(entity = HttpEntity(ContentTypes.`application/json`, JsObject().compactPrint))
+    val httpRequest = HttpRequest(entity = HttpEntity(ContentTypes.`application/json`, JsObject().compactPrint))
+    val poolingRestClient = new PoolingRestClient("https", "host", 443, 1, Some(testFlow(httpResponse, httpRequest)))
+    val request = mkJsonRequest(GET, Uri./, JsObject(), List.empty)
+
+    await(poolingRestClient.requestJson[JsObject](request)) shouldBe Right(JsObject())
+  }
+
+  it should "return a status code on request failure" in {
+    val httpResponse = HttpResponse(NotFound)
+    val httpRequest = HttpRequest(entity = HttpEntity(ContentTypes.`application/json`, JsObject().compactPrint))
+    val poolingRestClient = new PoolingRestClient("https", "host", 443, 1, Some(testFlow(httpResponse, httpRequest)))
+    val request = mkJsonRequest(GET, Uri./, JsObject(), List.empty)
+
+    await(poolingRestClient.requestJson[JsObject](request)) shouldBe Left(NotFound)
+  }
```

Comment:
Could you add a test on what happens if unmarshalling fails?

Reply:
Test added.